lib/screens/v2/screen_data.ex

```diff
@@ -133,7 +135,192 @@ defmodule Screens.V2.ScreenData do
+          integer() | nil,
+          DateTime.t()
+        ) ::
+          {Template.non_paged_layout(), non_paged_selected_instances_map()}
+  def resolve_paging(layout_and_instances, refresh_rate, now \\ DateTime.utc_now())
+
+  def resolve_paging(layout_and_instances, nil, _now) do
+    layout_and_instances
+  end
+
+  def resolve_paging({layout, instance_map}, refresh_rate, now) do
+    {unpaged_layout, selected_paged_slot_ids} = choose_pages(layout, refresh_rate, now)
+
+    # Now filter instance map by set of paged slot ids, then unpage
+    instance_map =
+      instance_map
+      |> Enum.filter(fn
+        {slot_id, _instance} when is_paged_slot_id(slot_id) ->
+          slot_id in selected_paged_slot_ids
+
+        _ ->
+          true
+      end)
+      |> Enum.map(fn {slot_id, instance} -> {Template.unpage(slot_id), instance} end)
+      |> Enum.into(%{})
+
+    {unpaged_layout, instance_map}
+  end
+
+  defp select_page_index(num_pages, refresh_rate, now) do
+    seconds_since_midnight = now.hour * 60 * 60 + now.minute * 60 + now.second
+    periods_since_midnight = div(seconds_since_midnight, refresh_rate)
+    rem(periods_since_midnight, num_pages)
+  end
+
+  @spec choose_pages(Template.layout(), integer(), DateTime.t()) ::
+          {Template.non_paged_layout(), MapSet.t(Template.paged_slot_id())}
+  defp choose_pages(layout, refresh_rate, now)
+
+  defp choose_pages(slot_id, _refresh_rate, _now) when is_non_paged_slot_id(slot_id) do
+    {slot_id, MapSet.new()}
+  end
+
+  defp choose_pages({slot_id, {layout_type, children}}, refresh_rate, now) do
+    {children, selected_paged_slot_ids} = choose_pages(children, refresh_rate, now)
+
+    {{slot_id, {layout_type, children}}, selected_paged_slot_ids}
+  end
+
+  defp choose_pages(layouts, refresh_rate, now) when is_list(layouts) do
+    selected_paged_slot_ids =
+      layouts
+      |> Enum.filter(fn
+        slot_id when is_paged_slot_id(slot_id) -> true
+        {slot_id, _} when is_paged_slot_id(slot_id) -> true
+        _ -> false
+      end)
+      |> Enum.group_by(
+        fn
+          slot_id when is_paged_slot_id(slot_id) -> Template.unpage(slot_id)
+          {slot_id, _} when is_paged_slot_id(slot_id) -> Template.unpage(slot_id)
+        end,
+        fn
+          slot_id when is_paged_slot_id(slot_id) -> Template.get_page(slot_id)
+          {slot_id, _} when is_paged_slot_id(slot_id) -> Template.get_page(slot_id)
+        end
+      )
+      |> Enum.map(fn {slot_id, page_indexes} -> {slot_id, Enum.max(page_indexes) + 1} end)
+      |> Enum.map(fn {slot_id, num_pages} ->
+        {select_page_index(num_pages, refresh_rate, now), slot_id}
+      end)
+
+    selected_layouts =
+      Enum.filter(layouts, fn
+        slot_id when is_paged_slot_id(slot_id) -> slot_id in selected_paged_slot_ids
+        {slot_id, _} when is_paged_slot_id(slot_id) -> slot_id in selected_paged_slot_ids
+        _ -> true
+      end)
```

Comment:
This might be a bit more readable if the helper functions were named (e.g. the first one could filter by `is_paged_layout`).